### PR TITLE
Create artifacts from URL

### LINKED ISSUE #1499: Support upload of schema definitons from URL
Discussion: https://github.com/Apicurio/apicurio-registry/discussions/1494
See also: https://github.com/Apicurio/apicurio-registry-content-sync-operator/issues/3

### PR DESCRIPTION
**Pull request by andreaTP**
Opening it as a draft to check if I'm following the guidelines of this repo ...

Resolves #1499

Question:
- should we extend this support to other endpoints? e.g. `updateArtifact` `createArtifactVersion`

### CODE REVIEW

**Comment by andreaTP** [comment]
Functionality is now covered by tests and CI is green, I would say it's ready for review.

**Comment by andreaTP** [comment]
Forgot to mention, that in this PR I'm also introducing the possibility of uploading an artifact specifying the content SHA, it's a common and [standard security rule](https://techdocs.akamai.com/download-ctr/docs/verify-checksum).
I expect it to be a desired feature when using the registry directly through the API and with automatic tools, probably not really affecting UI users.

**Comment by EricWittmann** [comment]
Sorry @andreaTP - I just noticed this when going through my backlog.  I'll have a look at this as soon as I can.  :)

**Comment by EricWittmann** [comment]
@smccarthy-ie ping on this PR.

File: app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java [hunk 1]
```diff
@@ -645,10 +663,58 @@ private ArtifactMetaData createArtifactWithRefs(String groupId, ArtifactType xRe
...
         }
+
+        // Mitigation for MITM attacks, verify that the artifact is the expected one
+        if (xRegistryArtifactSHA != null) {
+            String calculatedSha = Hashing.sha256().hashString(content.content(), StandardCharsets.UTF_8).toString();
+
+            if (!calculatedSha.equals(xRegistryArtifactSHA)) {
+                throw new BadRequestException("Provided Artifact SHA doesn't match with the content");
+            }
+        }
```
Comment (EricWittmann) [inline app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java:716]: Why does this address the MITM attack?  If the attacker can change the payload, can't they also just change the SHA as well?

Comment (andreaTP) [reply app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java:716]: A malicious actor can replace the server providing the file(especially if not using `https`), while the `sha` comes along with the request (and is not hosted on the remote hosting server).

Comment (EricWittmann) [reply app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java:716]: Ah I see - so useful in the `fromUrl` use-case more than in other cases.

**Comment by EricWittmann** [comment]
So I think we should reconsider the API changes you've made.  I don't think it's the right approach to pass the URL as a query parameter.  The content of the artifact should always be passed in the body of the operation (the `POST` in the case of artifact creation).  So I suggest, as an alternative, that we allow passing a URL when using `application/create.extended+json` as the body of the create.  Either as a new property or by re-using the existing `content` property.

image

Also, we should make sure any changes we make to the Registry API be first done in Apicurio Studio.  Or **at least** backported to the design in Apicurio Studio.  @andreaTP you may  not know exactly where that lives, so if not please ping me on gchat and I'll send you a link.

File: app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java (hunk 1 above)
Comment (andreaTP) [reply app/src/main/java/io/apicurio/registry/rest/v2/GroupsResourceImpl.java:716]: very correct 👍

**Comment by jsenko** [comment]
I agree with Eric's comment. A few additional thoughts:
- Should we perform the download in the request thread? Even if we limit the download size, this feels like a potential avenue for a DoS attack. Few ideas that could help mitigate that:
  - Global limit on number of concurrent downloads, either returning 429 or queuing.
  - Offload the download to a separate thread pool. This comes with a downside of having to deal with async operations, and there has to be a way for client to find out if the request succeeded.
- We should be specific about the type of the SHA hash. i.e. use `X-Registry-ArtifactSHA256` header name.
- In our REST client, it could compute the hash automatically.

**Comment by EricWittmann** [comment]
Interesting thoughts, thanks @jsenko .  I'll go in reverse order.  :)

+1 to `X-Registry-ContentSHA256` as the header name.  Alternatively we could have `X-Registry-Content-Hash` with a value that included both the algorithm **and** the hash value.  Or perhaps two headers - one for the hash value and one for the algorithm (default to SHA256).  That would be nice flexibility.

Not sure about moving the download logic to another threadpool.  I think a DoS attack is a reasonable concern.  But to mitigate that I think we would need to make the operation **itself** asynchronous (something we don't do right now at all).  I think this is what you're suggesting.  Instead, perhaps we could address this with client annotations to handle retries and timeouts.  In a multi-tenant deployment we could perhaps consider how to do rate limiting of these requests per-tenant?

**Comment by andreaTP** [comment]
@EricWittmann and @jsenko thanks for the feedback, very much appreciated!

I'll go with your proposals for everything but I think we need to better discuss the risk of DoS attacks.
I acknowledge the risk, but we might have some "higher level" mitigation against it (e.g. plain rate-limiting).
Going "async" using a separate TP, queuing etc. is completely doable but I'm worried about the user experience and the usability; do we have any longer-term vision around asynchronous actions for this project?
Before going down that path I would like to understand how a simple workflow like:
 - upload a first version of the API from URL
 - modify it (without waiting for it by polling)
 - get it back
 can look like in the case of async operations.

**Comment by andreaTP** [comment]
@EricWittmann @jsenko , sorry for the delay, I finally found the time to actually implement your suggestions.

Should be ready for another round of review 🙂 
If/when we agree on the details of the implementation I'll do the backporting wherever necessary.

**Comment by andreaTP** [comment]
and finally with a green CI 💚 😅

**Comment by EricWittmann** [comment]
By the way - I should have said I didn't think going async was a good idea.  I think in general my opinion is that async is not the right move, especially as Loom gets closer.  Also note that in RHOSR we already have per-tenant rate limiting that would prevent a denial of service attack from splashing over too hard on the other tenants.  So I think timeout handling should be sufficient.

Now I'm going to review the code again.  :)

File: app/src/test/java/io/apicurio/registry/noprofile/serde/ProtobufSerdeTest.java [hunk 2]
```diff
@@ -50,11 +50,12 @@
 public class ProtobufSerdeTest extends AbstractResourceTestBase {
 
     private RegistryClient restClient;
+    // Isolating this test in it's own groupId:
+    // io.apicurio.registry.rest.client.exception.ArtifactAlreadyExistsException: An artifact with ID 'google/protobuf/timestamp.proto' in group 'default' already exists.
+    private String groupId = "protobuf-serde-test";
```
Comment (EricWittmann) [inline app/src/test/java/io/apicurio/registry/noprofile/serde/ProtobufSerdeTest.java:55]: Nice.

File: app/src/test/java/io/apicurio/registry/noprofile/serde/ProtobufSerdeTest.java [hunk 3]
```diff
@@ -50,11 +50,12 @@
...
 
     @BeforeEach
-    public void createIsolatedClient() {
-        restClient = RegistryClientFactory.create(TestUtils.getRegistryV2ApiUrl());
-    }
+    public void createIsolatedClient() { restClient = RegistryClientFactory.create(TestUtils.getRegistryV2ApiUrl()); }
```
Comment (EricWittmann) [inline app/src/test/java/io/apicurio/registry/noprofile/serde/ProtobufSerdeTest.java:58]: Not necessary?  Maybe revert this...

File: client/src/main/java/io/apicurio/registry/rest/client/RegistryClient.java [hunk 4]
```diff
@@ -131,29 +131,29 @@ default ArtifactSearchResults listArtifactsInGroup(String groupId) {
...
     }
 
-    ArtifactMetaData createArtifact(String groupId, String artifactId, String version, ArtifactType artifactType, IfExists ifExists, Boolean canonical, String artifactName, String artifactDescription, String contentType, InputStream data);
+    ArtifactMetaData createArtifact(String groupId, String artifactId, String version, ArtifactType artifactType, IfExists ifExists, Boolean canonical, String artifactName, String artifactDescription, String contentType, String fromURL, String artifactSHA, InputStream data);
 
-    ArtifactMetaData createArtifact(String groupId, String artifactId, String version, ArtifactType artifactType, IfExists ifExists, Boolean canonical, String artifactName, String artifactDescription, String contentType, InputStream data, List<ArtifactReference> artifactReferences);
+    ArtifactMetaData createArtifact(String groupId, String artifactId, String version, ArtifactType artifactType, IfExists ifExists, Boolean canonical, String artifactName, String artifactDescription, String contentType, String fromURL, String artifactSHA, InputStream data, List<ArtifactReference> artifactReferences);
```
Comment (EricWittmann) [inline client/src/main/java/io/apicurio/registry/rest/client/RegistryClient.java:136]: Do we also need to allow passing the hashing algorithm here?

File: ui/src/app/pages/artifacts/components/uploadForm/uploadForm.tsx [hunk 5]
```diff
@@ -158,13 +160,46 @@ export class UploadArtifactForm extends PureComponent<UploadArtifactFormProps, U
...
+                        isRequired={false}
+                        type="text"
+                        id="form-sha"
+                        data-testid="form-sha"
+                        name="form-sha"
+                        aria-describedby="form-sha-helper"
+                        value={this.state.sha ? this.state.sha : ""}
+                        placeholder="SHA256 of the artifact"
+                        onChange={this.onShaChange}
                     />
```
Comment (EricWittmann) [inline ui/src/app/pages/artifacts/components/uploadForm/uploadForm.tsx:203]: This is good for now, but FYI I might change this UI experience to work more like what we have now in API Designer:

image

This way the user can validate the content in the browser before submitting it.  And we lessen the opportunities for users to send us URLs that might not respond quickly.

And it makes things more consistent across services.

Comment (EricWittmann) [reply ui/src/app/pages/artifacts/components/uploadForm/uploadForm.tsx:203]: Nothing to change for now on this.  :)

**Review by EricWittmann — CHANGES_REQUESTED** [review]
Just a few changes requested.  Looks good though!  Nice work.

File: common/src/main/resources/META-INF/openapi.json [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {
     "openapi": "3.0.2",
     "info": {
-        "title": "Apicurio Registry API [v2]",
+        "title": "[Andrea's playground] Apicurio Registry API [v2]",
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply common/src/main/resources/META-INF/openapi.json:4]: Argh 😓 sure

File: common/src/main/resources/META-INF/openapi.json [hunk 7]
```diff
@@ -2175,6 +2175,28 @@
                             "$ref": "#/components/schemas/EncodedArtifactName"
                         },
                         "in": "header"
+                    },
+                    {
+                        "name": "X-Registry-Content-Hash",
+                        "description": "Specifies the SHA-256 of the artifact to be verified.",
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply common/src/main/resources/META-INF/openapi.json:2181]: 👍

File: common/src/main/resources/META-INF/openapi.json [hunk 8]
```diff
@@ -2175,6 +2175,28 @@
...
+                    {
+                        "name": "X-Registry-Hash-Algorithm",
+                        "description": "The algorithm to use when checking the content validity. (available: SHA256,MD5; default: SHA256)",
+                        "schema": {
+                            "enum": [
+                                "SHA256",
+                                "MD5"
+                            ],
+                            "type": "string"
+                        },
```
[the remark this answers is not in the document]
Comment (andreaTP) [reply common/src/main/resources/META-INF/openapi.json:2197]: I used what is available for Jars ... I think we can add more when/if needed.
e.g.: https://repo1.maven.org/maven2/io/apicurio/apicurio-data-models/1.1.26/

File: client/src/main/java/io/apicurio/registry/rest/client/RegistryClient.java (hunk 4 above)
Comment (andreaTP) [reply client/src/main/java/io/apicurio/registry/rest/client/RegistryClient.java:136]: We can do it, of course, but I'm not sure how much this flexibility will be used in real world ...

File: ui/src/app/pages/artifacts/components/uploadForm/uploadForm.tsx (hunk 5 above)
Comment (andreaTP) [reply ui/src/app/pages/artifacts/components/uploadForm/uploadForm.tsx:203]: Nice 👍

**Comment by andreaTP** [comment]
Just connecting the dots together, you can review the UI running:

```
docker run --rm -it -p 8080:8080 ttl.sh/c75ca54628e298b57604acaf6e27b9855698b8e4/apicurio/apicurio-registry-mem:1d
```
🎉

**Pull request merged**